File: model_1/CS3233_2024_Midterm/false_sanctum_1_solution.py
import sys
sys.setrecursionlimit(300000)

class Node:
    __slots__ = ('size', 'lc', 'rc', 'lrun', 'rrun', 'dens')
    def __init__(self, size, lc, rc, lrun, rrun, dens):
        self.size = size
        self.lc = lc
        self.rc = rc
        self.lrun = lrun
        self.rrun = rrun
        self.dens = dens

def merge(L, R):
    if L.size == 0:
        return R
    if R.size == 0:
        return L
    size = L.size + R.size
    lc = L.lc
    rc = R.rc
    lrun = L.lrun
    if L.lrun == L.size and L.rc == R.lc:
        lrun = L.lrun + R.lrun
    rrun = R.rrun
    if R.rrun == R.size and L.rc == R.lc:
        rrun = R.rrun + L.rrun
    dens = L.dens + R.dens
    if L.rc == R.lc:
        dens = dens - (L.rrun * L.rrun) - (R.lrun * R.lrun) + ((L.rrun + R.lrun) * (L.rrun + R.lrun))
    return Node(size, lc, rc, lrun, rrun, dens)

def build_tree(node, l, r, tree, s):
    if l == r:
        tree[node] = Node(1, s[l], s[l], 1, 1, 1)
        return
    mid = (l + r) // 2
    left_child = 2 * node + 1
    right_child = 2 * node + 2
    build_tree(left_child, l, mid, tree, s)
    build_tree(right_child, mid+1, r, tree, s)
    tree[node] = merge(tree[left_child], tree[right_child])

def update_tree(node, l, r, idx, c, tree):
    if l == r:
        tree[node] = Node(1, c, c, 1, 1, 1)
        return
    mid = (l + r) // 2
    left_child = 2 * node + 1
    right_child = 2 * node + 2
    if idx <= mid:
        update_tree(left_child, l, mid, idx, c, tree)
    else:
        update_tree(right_child, mid+1, r, idx, c, tree)
    tree[node] = merge(tree[left_child], tree[right_child])

def query_tree(node, l, r, ql, qr, tree):
    if ql <= l and r <= qr:
        return tree[node]
    mid = (l + r) // 2
    left_node = None
    right_node = None
    if ql <= mid:
        left_node = query_tree(2*node+1, l, mid, ql, min(qr, mid), tree)
    if qr > mid:
        right_node = query_tree(2*node+2, mid+1, r, max(ql, mid+1), qr, tree)
    if left_node is not None and right_node is not None:
        return merge(left_node, right_node)
    elif left_node is not None:
        return left_node
    elif right_node is not None:
        return right_node
    else:
        return Node(0, None, None, 0, 0, 0)

def main():
    data = sys.stdin.read().splitlines()
    if not data: 
        return
    n, q = map(int, data[0].split())
    s = data[1].strip()
    tree = [None] * (4 * n)
    build_tree(0, 0, n-1, tree, s)
    out_lines = []
    index = 2
    for i in range(q):
        parts = data[index].split()
        index += 1
        if parts[0] == '1':
            pos = int(parts[1]) - 1
            c = parts[2]
            update_tree(0, 0, n-1, pos, c, tree)
        else:
            l = int(parts[1]) - 1
            r = int(parts[2]) - 1
            node = query_tree(0, 0, n-1, l, r, tree)
            out_lines.append(str(node.dens))
    sys.stdout.write("\n".join(out_lines))

if __name__ == '__main__':
    main()

File: model_1/CS3233_2024_Midterm/false_sanctum_1_solution.cpp
#include <iostream>
#include <string>
using namespace std;

const int MAXN = 200000;
struct Node {
    int size;
    char lc, rc;
    int lrun, rrun;
    long long dens;
    Node() : size(0), lc(0), rc(0), lrun(0), rrun(0), dens(0) {}
    Node(int sz, char left_char, char right_char, int left_run, int right_run, long long density)
        : size(sz), lc(left_char), rc(right_char), lrun(left_run), rrun(right_run), dens(density) {}
};

Node tree[4 * MAXN];
string s;

Node merge(Node L, Node R) {
    if (L.size == 0) return R;
    if (R.size == 0) return L;
    Node res;
    res.size = L.size + R.size;
    res.lc = L.lc;
    res.rc = R.rc;

    if (L.lrun == L.size && L.rc == R.lc) {
        res.lrun = L.lrun + R.lrun;
    } else {
        res.lrun = L.lrun;
    }

    if (R.rrun == R.size && L.rc == R.lc) {
        res.rrun = R.rrun + L.rrun;
    } else {
        res.rrun = R.rrun;
    }

    res.dens = L.dens + R.dens;
    if (L.rc == R.lc) {
        long long subtract = (long long)L.rrun * L.rrun + (long long)R.lrun * R.lrun;
        long long add = (long long)(L.rrun + R.lrun) * (L.rrun + R.lrun);
        res.dens = res.dens - subtract + add;
    }
    return res;
}

void build(int node, int l, int r) {
    if (l == r) {
        tree[node] = Node(1, s[l], s[l], 1, 1, 1);
        return;
    }
    int mid = (l + r) / 2;
    build(2 * node + 1, l, mid);
    build(2 * node + 2, mid + 1, r);
    tree[node] = merge(tree[2 * node + 1], tree[2 * node + 2]);
}

void update(int node, int l, int r, int idx, char c) {
    if (l == r) {
        s[l] = c;
        tree[node] = Node(1, c, c, 1, 1, 1);
        return;
    }
    int mid = (l + r) / 2;
    if (idx <= mid) {
        update(2 * node + 1, l, mid, idx, c);
    } else {
        update(2 * node + 2, mid + 1, r, idx, c);
    }
    tree[node] = merge(tree[2 * node + 1], tree[2 * node + 2]);
}

Node query(int node, int l, int r, int ql, int qr) {
    if (ql <= l && r <= qr) {
        return tree[node];
    }
    int mid = (l + r) / 2;
    Node left_node, right_node;
    bool left_used = false, right_used = false;
    if (ql <= mid) {
        left_node = query(2 * node + 1, l, mid, ql, min(qr, mid));
        left_used = true;
    }
    if (qr > mid) {
        right_node = query(2 * node + 2, mid + 1, r, max(ql, mid + 1), qr);
        right_used = true;
    }
    if (left_used && right_used) {
        return merge(left_node, right_node);
    } else if (left_used) {
        return left_node;
    } else if (right_used) {
        return right_node;
    }
    return Node();
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, Q;
    cin >> N >> Q;
    cin >> s;

    build(0, 0, N - 1);

    for (int q = 0; q < Q; q++) {
        int type;
        cin >> type;
        if (type == 1) {
            int i;
            char c;
            cin >> i >> c;
            update(0, 0, N - 1, i - 1, c);
        } else {
            int l, r;
            cin >> l >> r;
            Node res = query(0, 0, N - 1, l - 1, r - 1);
            cout << res.dens << '\n';
        }
    }

    return 0;
}